### ISSUE TITLE
Normalize geographical location names

### ISSUE THREAD

**Issue by vitorbaptista** [opentrials/opentrials#250]
OpenTrials allows the user to search for trials in a specific geographical location (e.g. UK, Brazil, ...). However, the location names aren't normalized. This means that searching for "UK" or searching for "United Kingdom" might return different results. This also creates issues when we're trying to get statistics around trials per geographical location.

To solve this, we need to normalize the names into a canonical form (as a first step). I found a few tools that might help ([here](https://github.com/Corollarium/geograpy2), [here](https://github.com/elyase/geotext), [here](http://chairnerd.seatgeek.com/fuzzywuzzy-fuzzy-string-matching-in-python/), and [here](http://www.geonames.org/)). Many look abandoned, so I'm not sure which is best.
# Non Goals
- We won't deal with corner cases (e.g. [here](http://explorer.opentrials.net/trials/0cea0d2b-3d5b-4003-8372-b26b1f5f40f2), [here](http://explorer.opentrials.net/trials/fd8cf6b9-053e-463c-b1e9-04c59e470135), [here](http://explorer.opentrials.net/trials/d37f1364-b3fd-4a00-a24a-b7b599ba29e4) and [here](http://explorer.opentrials.net/trials/96c41ab1-abf7-4ed8-b435-cc9d3e5221c6)), just log them;
- We won't deal with synonyms (e.g. "USA" and "United States");
- We won't deal with geographical "levels" (i.e. a filter for "Europe" returning trials in "Germany");
# Expected Output

A helper method/function/class (whatever makes sense) in the [processor](https://github.com/opentrials/processors) that returns the canonical name of a geographical location. It should log all normalizations it's doing, including if it was successful or not. If it's unable to figure it out, it should return the same name it received.
# Acceptance Criteria
- [ ] Every location in OpenTrials appear only once (there're only [~300 at the moment](http://api.opentrials.net/v1/search/autocomplete/location), so it's easy to check)

**Comment by roll** [comment]
@vitorbaptista why to do not deal with synonyms?

**Comment by vitorbaptista** [comment]
@roll Just to keep this issue (somewhat) small. We'll have to deal with synonyms later (specially for things like drug names), but we can create another issue for that.

**Comment by TertiumQuid** [comment]
Hi Vitor,

Just picking this up from email; for country locations the myTomorrows team just followed ISO 3166-1 and had no issues at all (except for a few edge cases, e.g. parsing countries in a comma-separated list which also had commas in the name). We don't allow free-text country searches so normalizing a location query is by definition not a problem for us.

Extracting country names from text is a different matter - IIRC all the trial registries publish country data in a parsable way, however data like sponsors and contacts often have location info in a free-text "address" field which you do need to parse. In my experience this is usually delimited somehow, and so we didn't need fancy logic here either.

**Comment by vitorbaptista** [comment]
@TertiumQuid thanks for the hints. How you handled corner cases like http://apps.who.int/trialsearch/Trial3.aspx?trialid=JPRN-UMIN000009528 (region defined as `Asia(except Japan)`) or http://apps.who.int/trialsearch/Trial3.aspx?trialid=JPRN-JapicCTI-121953 (region defined as `In about 30 countries around the world`)? I imagine in a case by case basis, right?

**Comment by roll** [comment]
I suppose logging successful normalisations will lead to a major noise even for debug log. But failed ones will be very useful to see in PaperTrails. Then we could hard-code map tables for unknown ones.

**Comment by roll** [comment]
Related datapackage - https://github.com/datasets/country-codes/blob/master/data/country-codes.csv

**Comment by Iararibeiro** [comment]
I agree with @roll, the corner cases are more related to don't have information/ or have vague information, than identify the country in the text. 
Existing some ways of dealing with that in my opinion and hard-code map table for that work, one suggestion to deal with that is using additional information  saying that its a single-country information or a region information.

**Comment by Iararibeiro** [comment]
I gonna add some information here just for future reference about the libraries suggested in the issue description.

- [geograpy2](https://github.com/Corollarium/geograpy2) = this library is not validate to download with pip and still brokes when use with npm, so you can download the project on github and copy the paste for the source. This package is cool because they use fuzzy to search.
- [geotext](https://github.com/elyase/geotext) = the most simple one, extract country names and codes from text (but only do this).
- [fuzzywuzzy](https://github.com/seatgeek/fuzzywuzzy) = make a fuzzy search in text, can be used as a helper with other library.
- [geonames](http://www.geonames.org/) = public database that covers a lot of countries and have a list of libs in different languages, they have two libs in python: 
                       1 - [geopy](code.google.com/p/geopy/) : cool one, but works more with address and zip codes, not with country names.
                       2 - [geonamesRdf](https://github.com/dsoprea/GeonamesRdf) : offers a lot of options to search places, countries, regions, and offers fuzzy search too. The problem is because all the search is made sending request to the geonames server, so sometimes can be slow and depends on the internet connection.

**Comment by roll** [comment]
@vitorbaptista 
It's `in development` but no one is assigned. So for cases like this may be responsible person (in this case Vitor) should be assigned.

**Comment by roll** [comment]
Have we tried to use http://nomenklatura.okfnlabs.org/ ?

**Comment by vitorbaptista** [comment]
Not for this, but we'll probably use it for more complex normalizations (e.g. organisation names)
